[Merged by Bors] - feat(algebra/non_unital_alg_hom): define non_unital_alg_hom

ocfnash opened the pull request:
The motivation is to be able to state the universal property for a magma algebra using bundled morphisms.

---
If we merge this, then I should be able to drop #6591 completely.

- [x] depends on: #7847

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(algebra/non_unital_alg_hom): define `non_unital_alg_hom`

The motivation is to be able to state the universal property
for a magma algebra using bundled morphisms.

ocfnash committed:
Add `to_non_unital_alg_hom`

ocfnash committed:
Generalise from `module` to just `distrib_mul_action`

ocfnash committed:
Golf and move `monoid_hom.inverse`

ocfnash committed:
Drop addition of `non_unital_alg_hom.lean`

The plan is to split this PR in two since the other changes are
deep and cause long build times.

ocfnash committed:
Revert "Drop addition of `non_unital_alg_hom.lean`"

This reverts commit 880e44bc0b8f11151f7df2f40b3540633c407a40.

ocfnash committed:
Merge branch 'master' into non_unital_alg_hom2

ocfnash committed:
Clean up after merge

ocfnash commented on src/algebra/non_unital_alg_hom.lean:
```diff
@@ -0,0 +1,182 @@
+{ .. (f : mul_hom A B).inverse g h₁ h₂,
+  .. (f : A →+[R] B).inverse g h₁ h₂ }
+
+@[simp] lemma coe_inverse (f : non_unital_alg_hom R A B) (g : B → A)
```
I had forgotten I wrote this `simp` lemma here.

The alternative approach would be to use `simps` when defining `inverse` (and add `to_fun := g`) but I think I actually prefer this since it's a more general lemma.

eric-wieser replied:
A possible downside of this simp lemma is that it applies before `map_add` has a chance to, a problem which the applied version doesn't share (I think)

ocfnash replied:
Good point though quick experiment indicates both possible lemmas suffer from this deficiency:
```lean
import algebra.non_unital_alg_hom

variables {R A B : Type*} [semiring R]
variables [non_unital_non_assoc_semiring A] [module R A]
variables [non_unital_non_assoc_semiring B] [module R B]
variables (f : non_unital_alg_hom R A B)
variables (g : B → A) (h₁ : function.left_inverse g f) (h₂ : function.right_inverse g f)

example (x y : B) : f.inverse g h₁ h₂ (x + y) = (g x) + (g y) :=
begin
  simp, -- applies only `non_unital_alg_hom.coe_inverse`
  -- Goal now: `g (x + y) = g x + g y`
  sorry,
end

-- Unnecessary as simp picks `inverse_apply` over `coe_inverse` below but maybe clarifies the point.
local attribute [-simp] non_unital_alg_hom.coe_inverse

@[simp] lemma non_unital_non_assoc_semiring.inverse_apply (x : B) :
  f.inverse g h₁ h₂ x = g x :=
rfl

example (x y : B) : f.inverse g h₁ h₂ (x + y) = (g x) + (g y) :=
begin
  simp, -- applies only `non_unital_alg_hom.inverse_apply`
  -- Goal now: `g (x + y) = g x + g y`
  sorry,
end
```

ocfnash committed:
Fix lemma name

I missed this when I generalised from extending linear_map to extending distrib_mul_action_hom

ocfnash committed:
Merge branch 'master' into non_unital_alg_hom2

ocfnash committed:
Relax typeclasses `module` to `distrib_mul_action`

Missed these in earlier work.

kbuzzard commented on src/algebra/non_unital_alg_hom.lean:
```diff
@@ -0,0 +1,182 @@
+multiplication, and scalar action. The multiplications are not assumed to be associative or
+unital, or even to be compatible with the scalar multiplications. In a typical application, the
+operations will satisfy compatibility conditions making them into algebras (albeit possibly
+non-associative and/or non-unital) but such conditions are not required to make this definition.
```
Maybe mention an example application?

ocfnash replied:
Good point; I've just pushed https://github.com/leanprover-community/mathlib/pull/7863/commits/76ec6cc51c28b694afc9016e27edb2b005f72914 which gives one immediate and one future example application.

ocfnash committed:
Improve module doc string, add examples

eric-wieser reviewed (approved):
bors d+

I think it would be worth adding a comment along the lines of ``TODO: add `non_unital_alg_equiv` when needed `` in the module docstring. I don't think it's necessary to actually add it in this PR though.

ocfnash committed:
Add comment about non_unital_alg_equiv
